src/Test.UnitTests.Sarif/Visitors/ExtractAllArtifactionLocationsVisitorTests.cs

```diff
@@ -0,0 +1,165 @@
+            }
+        }
+
+        [Fact]
+        public void ExtractAllArtifactLocationsVisitor_ExtractsMultipleLocationsSingleResult()
+        {
+            var sarifLog = new SarifLog
+            {
+                Runs = new[]
+                {
+                    new Run
+                    {
+                        Results = new[]
+                        {
+                            new Result
+                            {
+                                RuleId = TestConstants.RuleIds.Rule1,
+                                BaselineState = BaselineState.New,
+                                Locations = new []
+                                {
+                                    new Location
+                                    {
+                                        PhysicalLocation = new PhysicalLocation
+                                        {
+                                             ArtifactLocation = new ArtifactLocation
+                                             {
+                                                Uri = new Uri(TestConstants.FileLocations.Location1),
+                                             }
+                                        }
+                                    },
+                                    new Location
+                                    {
+                                        PhysicalLocation = new PhysicalLocation
+                                        {
+                                             ArtifactLocation = new ArtifactLocation
+                                             {
+                                                Uri = new Uri(TestConstants.FileLocations.Location2),
+                                             }
+                                        }
+                                    },
+                                    new Location
+                                    {
+                                        PhysicalLocation = new PhysicalLocation
+                                        {
+                                             ArtifactLocation = new ArtifactLocation
+                                             {
+                                                Uri = new Uri(TestConstants.FileLocations.Location3),
+                                             }
+                                        }
+                                    }
+                                }
+                            }
+                        }
+                    }
+                }
+            };
+
+            var visitor = new ExtractAllArtifactLocationsVisitor();
+            visitor.VisitSarifLog(sarifLog);
+
+            foreach (Location location in sarifLog.Runs[0].Results[0].Locations)
+            {
+                visitor.allArtifactLocations.Contains(location.PhysicalLocation.ArtifactLocation).Should().BeTrue();
+            }
+        }
```

Comment:
>} [](start = 8, length = 1)

Do you want to accumulate URIs that occur in other places, such as `Result.RelatedLocations`? That's what the code does today. If you don't want that, the visitor will need to maintain state ("I'm in `Result.Locations`!"). Again, it depends on your use case, and again, either way, you'll need to document the decision. Keep in mind that this is a general purpose SDK; what's right your your use case might not be appropriate in general.

Reply:
I tend to think the more useful SDK piece is to extract everything.  Downstream, the filer is going to need to take these extracted locations and filter on them in some way.  Given, however, that ArtifactLocation itself doesn't seem to contain any of the data necessary to make that decision as to where the node came from, it sounds like I may need to store that data.

I need to think on this a bit.  The filer definitely only cares about the locations extracted from the Results.Locations collection, as those are what we want to resolve into area paths for bugs.

I obviously can change the visitor to only hit the nodes I care about (probably at the Results.Location level), but I'm not sure that fits as well.

---
In reply to: [386742798](https://github.com/microsoft/sarif-sdk/pull/1794#discussion_r386742798) [](ancestors = 386742798)